lib/client.js

```diff
@@ -440,24 +442,25 @@ class Client extends EventEmitter {
     this[kRetryTimeout] = null
 
     const onDestroyed = (err) => {
-      for (const callback of this[kDestroyed]) {
+      const callbacks = this[kDestroyed]
+      this[kDestroyed] = null
+      for (const callback of callbacks) {
         callback(err, null)
       }
       cb(err, null)
     }
 
-    if (this.socket && !this.socket[kClosed]) {
-      let _err = null
+    if (!this.socket) {
+      process.nextTick(onDestroyed, null, null)
+    } else if (this.socket[kClosed] || this.socket[kError]) {
+      process.nextTick(onDestroyed, this.socket[kError], null)
+      this.socket[kError] = null
+    } else {
       this.socket
-        .on('error', (err) => {
-          _err = err
-        })
         .on('close', () => {
-          onDestroyed(_err)
+          onDestroyed(this.socket[kError])
         })
         .destroy(err)
-    } else {
-      process.nextTick(onDestroyed, null, null)
     }
```

Comment:
Was a problem here when `client.close` was invoked between `'error'` and `'close'` being emitted on the socket.